File: leetcode/0677-map-sum-pairs/solution.py
from typing import Dict


class TrieNode:
    def __init__(self):
        self.children: Dict[str, 'TrieNode'] = {}
        self.sum = 0


class MapSum:
    def __init__(self):
        self.root = TrieNode()
        self.key_values: Dict[str, int] = {}

    def insert(self, key: str, val: int) -> None:
        delta = val - self.key_values.get(key, 0)
        self.key_values[key] = val

        node = self.root
        node.sum += delta
        for ch in key:
            node = node.children.setdefault(ch, TrieNode())
            node.sum += delta

    def sum(self, prefix: str) -> int:
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return 0
            node = node.children[ch]
        return node.sum

File: leetcode/0677-map-sum-pairs/solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

class MapSum {
public:
    struct TrieNode {
        unordered_map<char, TrieNode*> children;
        int sum = 0;
    };

    TrieNode* root;
    unordered_map<string, int> keyValues;

    MapSum() {
        root = new TrieNode();
    }

    void insert(string key, int val) {
        int delta = val - (keyValues.count(key) ? keyValues[key] : 0);
        keyValues[key] = val;

        TrieNode* node = root;
        node->sum += delta;
        for (char ch : key) {
            if (!node->children.count(ch)) {
                node->children[ch] = new TrieNode();
            }
            node = node->children[ch];
            node->sum += delta;
        }
    }

    int sum(string prefix) {
        TrieNode* node = root;
        for (char ch : prefix) {
            if (!node->children.count(ch)) {
                return 0;
            }
            node = node->children[ch];
        }
        return node->sum;
    }
};
